### ISSUE TITLE
Window positions itself to the top left corner for a fraction of a second before going full screen

### ISSUE THREAD

**Issue by ValZapod** [clsid2/mpc-hc#128]
@clsid2 You can actually see it if you press Alt+Enter for a long time to understand what is the problem. The discussion in ffmpeg TRAC https://trac.ffmpeg.org/ticket/8072 : ffplay has somehow the same problem. They say it is absolutely normal, going fullscreen never was "smooth". I disagree. Really! Why one need this at all??? And in VLC as well!!!

**Comment by ValZapod** [comment]
@adipose Can you comment on it? Because D3D is also slow... But this is just bad.

**Comment by adipose** [comment]
@ValZapod 

I have never seen this behavior.  Maybe if the resolution changed before it fullscreened.

I'll take a look, but if going full screen the transition should be a low priority as long as the final result is clear, smooth fullscreen, imo.

**Comment by ValZapod** [comment]
@adipose It is not in d3d fullscreen, just in normal one. 

> press Alt+Enter for a long time

**Comment by adipose** [comment]
It's very difficult to control this sort of thing, as Windows handles the resizing of each window object. You can draw it correctly after resizing, but at first, Windows handles it (typically leaving the bits where they were).

I have been annoyed by this in the past and used various methods to get around (destroy window and recreate is one).  In this case it's not a good solution as that window is heavily integrated to the application. 

So what I did instead is create a shade over the screen while the window is resizing and redrawing.  If you think this is an improvement I may clean up the code and submit it for review.

https://mega.nz/#!dJwgQKLI!dpkbt5wOsePOfy-E_3tLre0i-i-5dWipu6e9-PXXCU0

The distortion you see is very minor, but I do dislike that the window moves its contents to an illogical place, for a second.  Having to hold alt-enter to appreciate it suggests it's not a big deal...but...

**Comment by ValZapod** [comment]
@adipose Yes! Good! 

> Having to hold alt-enter to appreciate

no, I saw it without that... The bug I think is not in windows but in SDL2 use. The same is in VLC and in ffplay!! And Pot player does not have this...
The only moment. It is now a little slow, because you now made the taskbar show itself somehow... (Try to hide taskbar with option hide taskbar while in destop mode.)

**Comment by adipose** [comment]
@ValZapod 

What evidence do you have it's caused by SDL, other than it happening in other applications?  The regular HWND often does the same thing.  If you hide a window, change the contents, and redisplay it somewhere else, it first blips the original contents.  It's quite ugly and annoying!

**Comment by ValZapod** [comment]
@adipose Should not there be a way to get rid off it with HWND? Also, I think your patch somehow breaks the highlighted window that is why taskbar is being redrawn.

**Comment by adipose** [comment]
@ValZapod 

Not sure what you are talking about with the taskbar.  I didn't see anything different about it.  Do you mean the windows taskbar?

**Comment by ValZapod** [comment]
Yes, windows taskbar. It looks like that you prioritize another window for a moment.

**Comment by adipose** [comment]
I do, but I don't see what you are talking about.

Try this build:

https://mega.nz/#!tZwWGS7Q!Er5jnngsmAmZHJIyeeKhV-XPF0dQ6IpPNV0gF5vvLY8

**Comment by ValZapod** [comment]
@adipose Now it is faster. Try to hide taskbar 
![aid563604-v4-728px-Hide-the-Windows-Taskbar-Step-2-Version-4 jpg](https://user-images.githubusercontent.com/31514790/71548764-00161f00-29c4-11ea-83dd-1a04fb1e4e7e.jpg)

**Comment by adipose** [comment]
I had tried that, but it made no difference.

Probably the cause is hiding the main window. I did it to prevent an ugly redraw but I will try something else.

**Comment by adipose** [comment]
Can you video the behavior with your phone?

**Comment by ValZapod** [comment]
Yep. First try to understand that Alt+Enter in windows Explorer will open file properties! And when you change the focus to next window that is what happens. Second, if you will have just only desktop (after Win-D) this will cause a mistake sound...

> Probably the cause is hiding the main window

Indeed!

**Comment by adipose** [comment]
I found a new api which has better results.  One side effect is, it is so fast, if you hold down alt-<enter>, you won't see the fullscreen at all sometimes.  But then, holding that down is dumb anyway ;)

https://mega.nz/#!AQoyHCRZ!Mlk4UI_y4ZOz4EuTdWkmZpgwsBH5fAvnwuq_sKFXLZ4

**Comment by ValZapod** [comment]
I have 4th generation of Intel (and even not Haswell refresh). So not so fast for me  (faster, yes, though) and the problem is still there. Also though it is faster I see that sometimes it stops working.

> you won't see the fullscreen at all sometimes

So not because it is fast I think.

**Comment by adipose** [comment]
@ValZapod 

Really...with this build I never see the problem.  Not even once.  I'm very surprised as the cloaking should prevent you from seeing anything.

Are you sure you tried the latest build?

**Comment by adipose** [comment]
Here's another build--I tried to force it to update, as perhaps your paint event is not erasing the window, or something.

https://mega.nz/#!tVhwWCxS!Cc7m_guW90cfD-MDF-hU7YHUunB3gFDWdxBcoQICZpY

**Comment by ValZapod** [comment]
Nope. Again the same. 

> your paint event is not erasing the window

Why that could be possible? Cache or something?

**Comment by adipose** [comment]
Try this build:

https://mega.nz/#!5M501SQL!LPhjYhndicQy-6by9iVCfHeWM50nR6HplL3pinJmL5o

It is supposed to force erase immediately.  If this doesn't work maybe the renderer is redrawing in between.

**Comment by ValZapod** [comment]
Again, bad. The only one that is working is the first https://mega.nz/#!dJwgQKLI!dpkbt5wOsePOfy-E_3tLre0i-i-5dWipu6e9-PXXCU0

BTW. I made a mistake. Pot Player also has the same issue. So...

**Comment by adipose** [comment]
Please send me a video of the behavior.

**Comment by adipose** [comment]
Try this build:

https://mega.nz/#!9JIWAYzQ!kkCWc3CnIm87b13aKX11OOVrZmNHHJ-KLW84bcGaSyk

Tweak this setting to something like 100 (maybe more for your slower computer).  This will give the window time to redraw before displaying. Hope it works!

![image](https://user-images.githubusercontent.com/3324395/71638607-4912e100-2c19-11ea-83f0-742c6a110600.png)

**Comment by ValZapod** [comment]
@adipose It becomes almost perfect after 1 (one). Every 10-15 alterations maybe it again happens...
WTF! Why on 100 it just does not work if you press Alt-Enter it just hides and does not work anymore)))))

And I have good computer, I have 2080 Ti))

**Comment by adipose** [comment]
Very strange, I can't reproduce that behavior.  I have used 0-500 and all work fine.  Here's another build with a couple changes--maybe they will help.

https://mega.nz/#!FE5DjQAK!GCFu515_Kuv8720Y2etojSD0h3QqIyJeFrwhwTXQJY8

**Comment by ValZapod** [comment]
@adipose Nothing changed. Maybe we can just use 1 as default, it is almost perfect. Sorry for such long answers, my gmail on android broke itself))) It does not give me pushes anymore.
Anyway, I think it parses Alt-Enter wrong.

**Comment by adipose** [comment]
If it causes issues for any delay, I can't release it.  1ms is about the same as no delay at all, which should work most of the time.  But if it fails at 100, it could fail at 1.  I have no idea why this would happen.

What does "un touch" it mean?  Are you saying you are holding down alt-enter and the window never pops up?  That could surely happen, as it repeatedly hides it, but what is the point in holding down that keystroke anyway?

**Comment by adipose** [comment]
Another test build here:

https://mega.nz/#!lNo3jABA!2Q175FK8k96FMk2zdr5PKjP9YSuBvVxfn8Wp3KiagpU

This just prevents the alt-enter from repeating too much while the delay is active.

**Comment by clsid2** [comment]
Don't waste to much time on this. Holding down a key is not normal usage, so any purely visual glitches are not really important imo.

**Comment by adipose** [comment]
@clsid2 Agreed.  But I consider it a regression over the behavior we have today, where you can see the window sizing back and forth.  Not that it's worth much to do so..., but at least there's some feedback that something is happening, in case you accidentally held it down.

**Comment by ValZapod** [comment]
@clsid2 @adipose The point here is that holding Alt-Enter should behave as double click, remember it is very fast. Yes latest build is perfect. Now 100, 500 (>=33) works nice. I suppose we can stop on 2. It is perfect on 2 (breaks maybe every 50 iterations if you hold it)... So, you said that API only supports Windows 10, we need #ifdef, right?

> 1ms is about the same as no delay at all, which should work most of the time

No, it surely behaves differently. I think that normal delay is less than 1 ms or maybe you only THINK it is 1 ms with that variable, while actually it has some overhead.

**Comment by adipose** [comment]
1ms is about the same as no delay at all, which should work most of the time

No, it surely behaves differently. I think that normal delay is less than 1 ms or maybe you only THINK it is 1 ms with that variable, while actually it has some overhead.

Allow me to rephrase.  It's about the same as a 0ms timer, which dumps the event to the end of the stack.  So any pending events run, then that event.  It still has an effect of delaying it vs. an immediate call.

But if you put 0ms, I don't call the timer.  So 1ms is the closest to an instant timer you can have. :)

**Comment by ValZapod** [comment]
Okay. But I mean we can stop on this. 2 already is perfect. You can of course set 50 or 100 but it is too slow as for me))

**Comment by adipose** [comment]
> So, you said that API only supports Windows 10, we need #ifdef, right?

No, I do a runtime check.  An ifdef would check it at compile time, which wouldn't allow it to work on multiple OS.

**Comment by ValZapod** [comment]
Yes, that is nice! Pull request?))

**Comment by adipose** [comment]
https://github.com/clsid2/mpc-hc/pull/206

**Comment by ValZapod** [comment]
@adipose Bravo! It is already in k-lite 15.3.3! Works perfectly (maybe every 50 iterations it breaks). Nice!

**Comment by clsid2** [comment]
Not everyone is happy with the new behavior. The transition to fullscreen can for example be weird/ugly when there is another window open behind MPC-HC with lots of white. Due to the transparency of MPC-HC that becomes visible for a short instant. Looks particularly bad when MPC-HC window is (nearly) maximized.

So I was thinking maybe to enable the cloaking only if FullscreenDelay>0, and use a delay only if FullscreenDelay>1.

What do you guys think?

**Comment by ValZapod** [comment]
@clsid2 First in windows 7 cloak does not work. Second, it was like that before this patch, I think.

**Comment by adipose** [comment]
No, the cloak hides the main window, which could make the window behind the main window appear briefly. Previously, that area would have always remained hidden.

The idea to only cloak if the delay is enabled is ok. I possibly could make it black out the area behind the main window, too.

**Comment by adipose** [comment]
https://github.com/clsid2/mpc-hc/pull/255

I think this is reasonable given that for best performance of the patch, a small delay is helpful anyway.

My old method of blacking out the screen before fullscreening would work, but since it takes some time to register the first paint event, I didn't like its effect on performance.
